Define DC/OS Development workflow.

orsenthil opened the pull request:
## High-level description

* https://jira.mesosphere.com/browse/DCOS-19716 - Document DCOS Pull Request Workflow

### Motivation

This document should give a high level overview of the DC/OS Development workflow, and assist the developers in understanding the requirements of a DC/OS Pull Request. The FAQ style of document should help clarify any common questions that a contributor might have.

Please ensure we have covered the common scenarios and the document is a useful reference for any contributor.

The [repo structure](https://github.com/dcos/dcos#repo-structure) and [pull request statuses](https://github.com/dcos/dcos#pull-requests-statuses) documents should be considered companion documents, and we should try to document more about the development workflow in suitable documents.  For e.g.  one document that I'd like to write (or see it written) is how to reproduce a test failure locally using dcos-e2e.

mesosphere-mergebot-dev commented:
This repo has @mesosphere-mergebot integration. You can interact with the following commands.
```
@mesosphere-mergebot merge-it  
@mesosphere-mergebot bump-ee  
@mesosphere-mergebot override-status  
@mesosphere-mergebot label [Ready For Review |Holding |Work In Progress |Ship It |Request For Comment] 
```

* PR creators can apply one of  `[Ready For Review |Work In Progress]`. [Owners](https://github.com/dcos/dcos/blob/master/owners.json) can apply any label.

orsenthil commented:
@mesosphere-mergebot label ready for review

orsenthil commented:
@mesosphere-mergebot bump-ee

orsenthil commented:
@adamtheturtle : I'd appreciate your review on this doc.

adamtheturtle commented:
@orsenthil I will review this tomorrow 😄

orsenthil commented:
@urbanserj - Good thinking. I will minimize the references to the Enterprise repo, and go with the assumption that this document addresses both Mesosphere contributors and community contributors.

adamtheturtle commented:
As "workflow" is one word there is no need for the hyphen in the filename.

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+1. Open Source DC/OS (https://github.com/dcos/dcos)
+2. Closed Source Mesosphere Enterprise DC/OS (https://github.com/mesosphere/dcos-enterprise)
+
+In our design, the closed source Mesosphere Enterprise DC/OS depends on the Open Source DC/OS, [builds on top of the artifact](https://github.com/mesosphere/dcos-enterprise/blob/master/util/build.py) from the Open source DC/OS.
```
❓  "In our design, the closed source" > "Closed Source".

"Open source DC/OS" > "Open Source DC/OS".

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+
+#### How do I contribute a change to dcos/dcos ?
+
+If a change is only to Enterprise DC/OS, the developer should create the pull request against the Enterprise Repository.
```
This is not an answer to the above question, which is confusing.

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+Mesosphere DC/OS has two primary repositories.
+
+1. Open Source DC/OS (https://github.com/dcos/dcos)
+2. Closed Source Mesosphere Enterprise DC/OS (https://github.com/mesosphere/dcos-enterprise)
```
The name changed at some point from Enterprise DC/OS to DC/OS Enterprise. Please change that throughout this document.

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+
+#### What should be my target branch?
+
+All Pull Requests should be against `master` branch, and the bug-fixes should be actively backported to the
```
This is confusing / not right. It is not that all PRs should be against `master` but that all *changes* should be against `master`.

adamtheturtle replied:
Is that even the case? Perhaps we will want to change something in an older version e.g. in a feature which has been removed in `master`.

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+
+All Pull Requests should be against `master` branch, and the bug-fixes should be actively backported to the
+supported versions of DC/OS. The dcos/dcos repo will have the list of actively supported branches, and developers
+should create the back-port pull requests if they have to include the change in the particular version DC/OS.
```
"the particular version DC/OS" "the particular version *of* DC/OS".

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+
+#### What should I do after I create a pull request against DC/OS?
+
+We use labels, to set the status of a pull-request. All pull requests start at "Work In
```
❓ We have "pull-request", "pull request" and "Pull Request" in this document.

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+The above command will set the label of the PR to "Ready For Review" and it is crucial,
+as it will now fall under the radar of Mergebot to monitor for status updates to the PR.
+
+And also, do update to Enterprise reference using the command:
```
"do update to Enterprise reference" - I'm not sure exactly what you mean by this. But I imagine something like.

"Each Pull Request to Open Source DC/OS is matched by a Pull Request to DC/OS Enterprise. Use the following command to create or update that matching Pull Request:"

adamtheturtle commented on docs/work-flow.md:
````diff
@@ -0,0 +1,176 @@
+@mesosphere-mergebot bump-ee
+```
+
+The will create a Pull Request against the dcos-enterprise repository and share a link to the pull-request in the github comments.
````
"The will create" > "This will create".

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+* Owners have a responsibility to review the automated train PRs, and help
+  developers with review for PRs against dcos/dcos, mesosphere/dcos-enterprise.
+
+* Owners also have the ability to manually add "ship-it" label to a PR to include it a merge train. This can
```
'to manually add "ship-it" label to a PR' > '**a** "ship-it" label'

adamtheturtle commented on docs/work-flow.md:
````diff
@@ -0,0 +1,176 @@
+git push origin mergebot/dcos/1.11/42
+```
+
+b. If you are an open source contributor, no EE contributions is required; the Bump-EE PR only
````
> contributions is required

Either "contribution" or "are".

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+
+TODO (skumaran) : DCOS_OSS-1906 - Github CODEOWNERS for dcos/dcos
+
+We expect two approvals at the last commit of your PR (HEAD of your branch), and
```
Instead of "we expect", let's be clear and say something like "Mergebot requires".

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+
+It is an indicator that your PR is safe to merge, and mesosphere-mergebot applies a "Ship It" label to those PRs.
+
+#### When does my PR get's merged, then?
```
"get's" > "get"

adamtheturtle commented:
> how to reproduce a test failure locally using dcos-e2e.

We have [a Wiki article](https://wiki.mesosphere.com/pages/viewpage.action?pageId=117081847) which covers most of this and I will bring it to these documents - ping me sometime and we can discuss doing this.

adamtheturtle commented on docs/work-flow.md:
```diff
@@ -0,0 +1,176 @@
+
+The will create a Pull Request against the dcos-enterprise repository and share a link to the pull-request in the github comments.
+
+a. If your changes require Enterprise only addition, you are expected to update the pull request thus created, with your changes.
```
"addition" > "additions" (or "changes")

mesosphere-mergebot-dev commented:
:wave: This PR has been inactive for 3 days. Please review the status checks to see what needs to be done to move it forward or change the label to Work in Progress. Thank you!

orsenthil commented:
@mesosphere-mergebot label Work In Progress

orsenthil commented:
Hi @urbanserj  - I considered separating this into two documents, but in the end I realized that we need a single document that gives complete overview of Pull Request life cycle.  Since Enterprise repo was referenced multiple times, and this documented was aimed at "Mesosphere" contributors, I thought, it is better to include in `mesosphere/dcos-enterprise` repo. Here is the https://github.com/mesosphere/dcos-enterprise/pull/2838 that supersedes this one. 
An update to `contribution.md` can written separately that will focus only on Open Source contributors.

---- 

@adamtheturtle  - Thank you for your review comments. I realized that this PR was not very cohesive, was not consistent in usage of the terms. I have addressed that mistake and I changed the format of the document from FAQ style to a process overview style with sections and explanations. As I explained to Sergey above, since this references `dcos-enterprise` multiple times, I felt this document should live in `dcos-enterprise` repo. I have addressed your all review comments in this modified document. Please continue with  your review here - https://github.com/mesosphere/dcos-enterprise/pull/2838

Thank you!

orsenthil commented:
Superseded by https://github.com/mesosphere/dcos-enterprise/pull/2838
